vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1179 @@
+                        shorts_.emplace_back(up_up, e_rr_type::CHANY, loc_chan_idx[std::make_tuple(up_up, e_rr_type::CHANY)],
+                                             loc, e_rr_type::CHANY, loc_chan_idx[std::make_tuple(loc, e_rr_type::CHANY)]);
+                }
+                if (vertex->links[2] != nullptr && vertex->links[2]->visited == vertex) {
+                    if (loc_chan_idx.find(std::make_tuple(right, e_rr_type::CHANX)) != loc_chan_idx.end())
+                        shorts_.emplace_back(right, e_rr_type::CHANX, loc_chan_idx[std::make_tuple(right, e_rr_type::CHANX)],
+                                             loc, e_rr_type::CHANY, loc_chan_idx[std::make_tuple(loc, e_rr_type::CHANY)]);
+                }
+                if (vertex->links[3] != nullptr && vertex->links[3]->visited == vertex) {
+                    if (loc_chan_idx.find(std::make_tuple(down, e_rr_type::CHANY)) != loc_chan_idx.end())
+                        shorts_.emplace_back(down, e_rr_type::CHANY, loc_chan_idx[std::make_tuple(down, e_rr_type::CHANY)],
+                                             loc, e_rr_type::CHANY, loc_chan_idx[std::make_tuple(loc, e_rr_type::CHANY)]);
+                }
+            }
+        }
+    }
+
+    /*
+     * Remove a graph represented by a root node.
+     * Clean up of earlier stages
+     */
+    void delete_nodes(node_* root) {
+        std::queue<node_*> walker;
+        std::stack<node_*> back_walker;
+        walker.emplace(root);
+        while (!walker.empty()) {
+            node_* vertex = walker.front();
+            walker.pop();
+            back_walker.emplace(vertex);
+            for (int i = 0; i < 4; ++i) {
+                if (vertex->links[i] != nullptr) {
+                    if (vertex->links[i]->visited == vertex) {
+                        walker.emplace(vertex->links[i]);
+                    }
+                }
+            }
+        }
+        while (!back_walker.empty()) {
+            node_* vertex = back_walker.top();
+            back_walker.pop();
+            for (int i = 0; i < 4; ++i) {
+                node_* temp = nullptr;
+                if (vertex->links[i] != nullptr) {
+                    temp = vertex->links[i];
+                    temp->links[(i + 2) % 4] = nullptr;
+                }
+            }
+            delete (vertex);
+        }
+    }
+
+    /*
+     * Process FPGA Interchange nodes
+     */
+    void process_nodes() {
+        auto wires = ar_.getWires();
+        for (auto const& node : ar_.getNodes()) {
+            std::tuple<int, int> base_wire_{wires[node.getWires()[0]].getTile(), wires[node.getWires()[0]].getWire()};
+            int node_id = wire_to_node_[base_wire_];
+            std::set<location> all_possible_tiles = node_to_locs_[node_id];
+            int root_x, root_y;
+            std::tie(root_x, root_y) = *all_possible_tiles.begin();
+            std::set<location> empties;
```

Comment:
Worth adding a comment explaining what empties is and why they are needed, just for future reference.